services/catalog/src/simcore_service_catalog/services/registry.py

```diff
@@ -0,0 +1,78 @@
+""" NOTE that this not refer ONLY to the docker registry but rather access to all published services both
+    - as images in the docker-registry
+    - as code in function-services
+"""
+
+import logging
+from typing import Any, TypeAlias, cast
+
+from models_library.function_services_catalog.api import iter_service_docker_data
+from models_library.services_metadata_published import ServiceMetaDataPublished
+from models_library.services_types import ServiceKey, ServiceVersion
+from pydantic import ValidationError
+
+from .director import DirectorApi
+from .function_services import get_function_service_as_model, is_function_service
+
+_logger = logging.getLogger(__name__)
+
+
+ServiceMetaDataPublishedMap: TypeAlias = dict[
+    tuple[ServiceKey, ServiceVersion], ServiceMetaDataPublished
+]
+
+
+_error_already_logged: set[tuple[str, str]] = set()
+
+
+async def get_registered_services_map(
+    director_client: DirectorApi,
+) -> ServiceMetaDataPublishedMap:
+    """Lists all services registered either in code (functional services) or the docker registry"""
+    registry_services = cast(
+        list[dict[str, Any]], await director_client.get("/services")
+    )
+
+    # NOTE: functional-services are services w/o associated image
+    services: ServiceMetaDataPublishedMap = {
+        (s.key, s.version): s for s in iter_service_docker_data()
+    }
+    for service in registry_services:
+        try:
+            service_data = ServiceMetaDataPublished.parse_obj(service)
+            services[(service_data.key, service_data.version)] = service_data
+
+        except ValidationError:  # noqa: PERF203
+            # NOTE: this is necessary until director API response does NOT provides any guarantee
```

Comment:
Maybe 
```python
# NOTE: this is necessary until director API response DOES provide some guarantees
```
? 🤔

Reply:
of course since some of the images are 10 years old... and some people were using fake emails at some point... :rofl:

Reply:
Ok, I get that. My point was more the comment. I don't understand it. The comment I wrote above seems to make more sense to me. But maybe I am just not enough into the details...